### PR TITLE
Add a Vue app for displaying collected data in charts

### PR DESCRIPTION
**Pull request by sknep**
This rather large PR adds a USWDS-skinned VueJS 3 app that displays library data from the backend in ChartJS-powered data visualizations with a router and unit test coverage. Addresses, in the simplest way possible, the stories captured in #103, #104 , #105. Recommend closing those tickets when merging this, and creating new tickets for follow-up work.

You should be able to view data for one of the sample library datasets for any day during May 2022.

Note: Some backend work to support this front-end has also been merged into this branch. 

Please pull the branch and run `docker compose build` on the /imls-frontend followed by `docker compose up` on both the /imls-frontend and /imls-backend folders. You may need to run an initial backend migration; instructions for that are available in the [backend Readme](https://github.com/IMLS/estimating-wifi/blob/7185ba26c8fa5a00360b1aae3a9b687475cc405e/imls-backend/README.md)

Add-Vue-App-151

### CODE REVIEW

**Comment by sknep** [comment]
@meenubhooshanan can you help with the conflicts on the backend files? i want to make sure that stuff is correct.

File: imls-frontend/src/store/store.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { reactive, computed, readonly } from "vue";
+
+const BASEURL = "http://127.0.0.1:3000";
```
[the remark this answers is not in the document]
Comment (sknep) [reply imls-frontend/src/store/store.js]: addressed

File: imls-frontend/test/setup.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+const knownGoodDevicesPerHourMock1 = [0,1,2,3,4];
+const knownGoodDevicesPerHourMock2 = [9,9,9,9,9];
+const knownEmptyDevicesPerHourMock = [0,0,0,0,0];
+const errorMock = {
+  details: "unexpected \"A\" expecting delimiter (.)",
+  message: "\"failed to parse filter (notARealID)\" (line 1, column 4)"
+}
+
+export const restHandlers = [
+  rest.get('http://127.0.0.1:3000/rpc/bin_devices_per_hour', (req, res, ctx) => {
```
[the remark this answers is not in the document]
Comment (sknep) [reply imls-frontend/test/setup.js]: addressed

File: imls-frontend/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
+FROM node:18-buster-slim
```
[the remark this answers is not in the document]
Comment (sknep) [reply imls-frontend/Dockerfile:1]: @wesley-dean-gsa -- is this something I should dismiss, or would you like us to revise this dockerfile?

File: imls-frontend/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+WORKDIR /app
+ENV NODE_ENV=development
+COPY . /app/
+RUN npm install
+
+# This CMD will attempt to make sure the node_modules are
+# present which may be problematic if there is no connection
+# to the Internet (npm install will hang); therefore, in
+# such a use-case, specify a CMD with 'npm run dev' instead.
+CMD ["/bin/sh -c 'npm install && npm run dev'"]
```
[the remark this answers is not in the document]
Comment (sknep) [reply imls-frontend/Dockerfile:14]: @wesley-dean-gsa -- what should I do about this? Want to help?

File: imls-frontend/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (wesley-dean-gsa) [reply imls-frontend/Dockerfile:1]: @sknep yeah, we can discuss.  Not a problem.

File: imls-frontend/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (wesley-dean-gsa) [reply imls-frontend/Dockerfile:14]: @sknep happy to help.  This is related, by the way, to the [other finding](https://github.com/IMLS/estimating-wifi/pull/151#discussion_r951748002).

**Comment by wesley-dean-gsa** [comment]
Also, @sknep, insert "giddy" emoji here!

**Comment by wesley-dean-gsa** [comment]
@sknep Ok.  Reviewed the Dockerfile.

Yes, it is correct that there is no `USER` statement and this means that, by default, the image is run as the user `root` (uid 0) and that can be a problem.  Similarly, there is no `RUN adduser` command to add an account one could use in conjunction with a `USER` statement.

Suppose we add a user, `pi` (or whatever), to the image via `RUN adduser pi` and then tell it to use that account via `USER pi`.  Cool.  That handles the security concern.

If one runs a container based on that image while bind-mounting the local directory (i.e., for hot reloads), then new files (e.g., those in node_modules) are written as the `pi` user which probably doesn't translate directly to the developer's account on the local system.  As a result, stuff in node_modules is owned by a separate account so deleting those files, overwriting them, etc. results in a permissions error for the local developer.

One way around this is to start the container by using docker's `-u` option like this:

```shell
docker run -u "$(id -u)" ....
```

That tells docker to run the container as if the image had been built with a `USER` statement specifying the local developer's account.  That is, we get the protection of running as a non-root user while avoiding the permissions issues associated with using a separate user account.  However, this still trips the `USER`-related warnings we're seeing.

Another option is to pass along the UID of the local developer as a parameter and use a startup script that, as a last step, runs `chown -R` to fix the ownership on the files in the directory.  This is also problematic because the script will need to use elevated privileges to run `chown` (typically via `sudo`) which means that a problematic script could do stuff as root which is the exact problem we're trying to avoid.

Yet another option is to configure directories that will be updated _by node, not the developer_ (e.g., node_modules) as a separate named volume.  This gets around the file writing / permissions concern while also resolving the `USER` security concerns.  However, this will also mean that the developer running the image would need to remember to attach the volume(s):

```shell
docker volume create "volumename"  # only needs to be done the first time
docker run -v "volumename:/app/node_modules" ...
```

That'll work, but it's something else for the developer to remember.  It's possible to codify this in a docker-compose.yml file and have the developer run `docker-compose up` instead of invoking `docker` directly (unless they wanted to handle the arguments and such themselves).  This would require that the developer has docker-compose installed as well as the supporting Python tooling (because docker-compose is written in Python).

All of this boils down to the tradeoff associated with allowing hot reloads.  We could make all of the concerns go away instantly by doing away with hot reloads; however, this would require a rebuild of the image for every single change a developer makes locally before it becomes visible.

**Comment by sknep** [comment]
> All of this boils down to the tradeoff associated with allowing hot reloads. We could make all of the concerns go away instantly by doing away with hot reloads; 

I prefer working in node locally to building in docker anyways, so as long as I can still `npm i` and then `npm run dev` or `npm run test`, i'll be happy. If we just use docker for checking builds in CI/CD stuff, and not for real local development, is that ok?

**Comment by cantsin** [comment]
> I prefer working in node locally to building in docker anyways

sgtm. i think it's reasonable to expect front end devs to be able to work locally? hot reloading in docker has always been iffy so i'm more than happy to drop this "feature".

**Review by cantsin — APPROVED** [review]
LGTM! excited to see this go in!!

**Review by jadudm — APPROVED** [review]
Having seen the work in progress, and a review of the PR, this looks good. MERGE!

**Pull request merged**